projects/com.oracle.truffle.llvm.parser.base/src/com/oracle/truffle/llvm/parser/base/model/LLVMToBitcodeAdapter.java

```diff
@@ -314,6 +317,28 @@ public static ResolvedType unresolveType(VectorType type) {
         return new ResolvedVectorType(type.getLength(), elementType);
     }
 
+    public static FunctionDeclaration resolveFunctionDef(LLVMParserRuntime runtime, FunctionDef def) {
+        return resolveFunctionHeader(runtime, def.getHeader());
+    }
+
+    public static FunctionDeclaration resolveFunctionHeader(LLVMParserRuntime runtime, FunctionHeader header) {
+        Type returnType = resolveType(runtime.resolve(header.getRettype()));
+        List<Type> args = new ArrayList<>();
+        boolean hasVararg = false;
+        for (Parameter arg : header.getParameters().getParameters()) {
+            assert !hasVararg; // should be the last element of the parameterlist
+            if (runtime.resolve(arg.getType().getType()).isVararg()) {
+                hasVararg = true;
+            } else {
+                args.add(resolveType(runtime.resolve(arg.getType().getType())));
+            }
+        }
+        FunctionType funcType = new FunctionType(returnType, args.toArray(new Type[args.size()]), hasVararg);
+        FunctionDeclaration funcDecl = new FunctionDeclaration(funcType);
+        funcDecl.setName(header.getName().substring(1));
```

Comment:
Making FunctionType implement ValueSymbol would remove the need to create a FunctionDeclaration here.